app/assets/javascripts/libfigure/define-figure.js

```diff
@@ -81,23 +80,46 @@ var progressionString = "to new neighbors";
 
 // ================
 
-function parameter_strings(move, parameter_values) {
+function parameter_strings(move, parameter_values, prefs) {
+  if (!prefs) {throw_up('forgot prefs argument to parameter_strings');}
   // new! improved! catch null and undefined, and convert them to '____', without calling the individual parameter function
   var formal_parameters = parameters(move);
   var acc = [];
   for (var i=0; i<parameter_values.length; i++) {
     var pvi = parameter_values[i];
+    var term;
     if ((pvi === undefined) || (pvi === null)) {
-      acc.push('____');
+      term = '____';
     } else if (formal_parameters[i].string) {
-      acc.push(formal_parameters[i].string(pvi,move));
+      term = formal_parameters[i].string(pvi,move);
     } else {
-      acc.push(String(pvi));
+      term = String(pvi);
     }
+    acc.push(preferenceParameter(prefs,formal_parameters[i],term));
   }
   return acc;
 }
 
+// called when we don't know if the parameter is a dancer
+function preferenceParameter(prefs, formal_parameter, actual_parameter) {
+  var term = actual_parameter;
+  if (formalParamIsDancers(formal_parameter) && (term in prefs.dancers)) {
+    return prefs.dancers[term];
+  } else {
+    return term;
+  }
+}
+
+// called when we do know the parameter is a dancer
+function preferenceDancers(prefs, actual_parameter) {
+  var term = actual_parameter;
+  if (term in prefs.dancers) {
+    return prefs.dancers[term];
+  } else {
+    return term;
+  }
+}
+
```

Comment:
what's the move-equivalent of this function? Shouldn't it be named consistently and appear near here sourcecodewise?

Reply:
`moveSubstitution` is the equivalent. 
We should rename `preferenceDancers` to `dancerSubstitution`.